### PR TITLE
fix: use the appropriate package manager when initializing a wrangler project

### LINKED ISSUE #353: wrangler init should support yarn
On `wrangler init`, if it detects that `yarn` is present, it should prompt the user and ask if they wish to continue using `yarn` (instead of `npm`)

### PR DESCRIPTION
**Pull request by petebacondarwin**
Previously, when we initialized a project using `wrangler init`, we always used npm as the package manager.

Now we check to see whether npm and yarn are actually installed, and also whether there is already a lock file in place before choosing which package manager to use.

Fixes #353

**Notes**:
The following features "could" be added later:
- ask the user which package manager they want to use, if it is not clear from the binaries and lock files
- allow the user to force which package manager to use via command line arguments

### CODE REVIEW

File: packages/wrangler/src/package-manager.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import { existsSync } from "node:fs";
+import { join } from "node:path";
+import { execa, execaCommandSync } from "execa";
+
+export interface PackageManager {
```
Comment (threepointone) [inline packages/wrangler/src/package-manager.ts:5]: This doesn't need to be exported.

File: packages/wrangler/src/package-manager.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+function supportsNpm(): Promise<boolean> {
+  return supports("npm");
+}
+
+/**
+ * Manage packages using npm
+ */
+export class NpmPackageManager implements PackageManager {
```
Comment (threepointone) [inline packages/wrangler/src/package-manager.ts:48]: I don't feel strongly about this, but why is this a class? Can it be a regular object like before?

**Comment by petebacondarwin** [comment]
I am actually flabbergasted that this passed CI first time! BOOM!

File: packages/wrangler/src/__tests__/index.test.ts [hunk 3]
```diff
@@ -404,3 +345,187 @@ describe("wrangler", () => {
     });
   });
 });
+
+/**
+ * This helper function will run a bunch of tests parameterized by the npm and yarn setup.
+ */
+function initAndInstall(
```
Comment (threepointone) [inline packages/wrangler/src/__tests__/index.test.ts:352]: It wasn't obvious that this function actually runs tests. Could you extract the tests into another function (say, `testInitCommand`, that accepts a `PackageManagerProtoType` arg), and have this function accept that function as an arg instead? So up above, you'd have, for example  - 



```js
function testInitCommand(PackageManagerPrototype){
  it("should create a package.json, and install packages, if none is found and user confirms", async () => {
        mockConfirm(
          {
            text: "No package.json found. Would you like to create one?",
            result: true,
          },
          {
            text: "Would you like to use typescript?",
            result: false,
          }
        );
       // ... and so on


initAndInstall(
      { npm: false, yarn: true, npmLockFile: true, yarnLockFile: true },
      YarnPackageManager.prototype,
      testInitCommand
)
```

File: packages/wrangler/src/package-manager.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
+import { existsSync } from "node:fs";
+import { join } from "node:path";
+import { execa, execaCommandSync } from "execa";
+
+export interface PackageManager {
+  addDevDeps(...packages: string[]): Promise<void>;
+  install(): Promise<void>;
+}
+
+export async function getPackageManager(root: string): Promise<PackageManager> {
```
Comment (threepointone) [inline packages/wrangler/src/package-manager.ts:10]: 100% bet you're going to get people asking for pnpm once they realise we're offering this option. We can do it later of course, but it's going to happen. (especially with folks slowly moving away from yarn)

**Review by threepointone — APPROVED** [review]
It's looking fine, hope you've tested it manually! Some notes if you'd like to consider

**Comment by threepointone** [comment]
(woke up with jaw pain and waiting for painkillers to kick in)

Thinking about this some more: considering we mock the actual binaries, running every combination of the package manager situations with every `init` test isn't really testing much, but does make the suite more complicated to read and write. Do you suppose  we should instead just write a few tests for each of those combinations that test which package manager gets selected, and keep that separate from the init tests?

File: packages/wrangler/src/package-manager.ts (hunk 2 above)
Comment (petebacondarwin) [reply packages/wrangler/src/package-manager.ts:48]: Coming from an OO background, a class seemed like the obvious paradigm here.
I am interesting in the argument against using a class for this sort of thing?

**Comment by petebacondarwin** [comment]
> (woke up with jaw pain and waiting for painkillers to kick in)
> 
> Thinking about this some more: considering we mock the actual binaries, running every combination of the package manager situations with every `init` test isn't really testing much, but does make the suite more complicated to read and write. Do you suppose we should instead just write a few tests for each of those combinations that test which package manager gets selected, and keep that separate from the init tests?

Yes - I think that this makes a lot of sense. While "integration" tests are valuable, here we just need to unit test the bit of the code that selects the correct package manager. I'll update the tests.

File: packages/wrangler/src/package-manager.ts (hunk 2 above)
Comment (threepointone) [reply packages/wrangler/src/package-manager.ts:48]: There's no state, no constructor arguments, and there will only be one instance/singleton required. So there's no visible benefit of using a class here (unless I'm missing something? Does it make types or tests simpler? I could be wrong ofc.)

Comment (petebacondarwin) [reply packages/wrangler/src/package-manager.ts:48]: Yes - all good points. And in fact on reflection, using a class actually makes the testing slightly more awkward (see `...prototype` bits. I'll change.

File: packages/wrangler/src/package-manager.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/wrangler/src/package-manager.ts:5]: It does now for my new tests 😄

File: packages/wrangler/src/__tests__/index.test.ts (hunk 3 above)
Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/index.test.ts]: These tests and this function are now gone 💀

File: packages/wrangler/src/package-manager.ts (hunk 4 above)
Comment (petebacondarwin) [reply packages/wrangler/src/package-manager.ts:11]: 🤫

File: packages/wrangler/src/package-manager.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  const [hasYarn, hasNpm] = await Promise.all([supportsYarn(), supportsNpm()]);
+  const hasYarnLock = existsSync(join(root, "yarn.lock"));
+  const hasNpmLock = existsSync(join(root, "package-lock.json"));
+
+  if (hasYarn && hasYarnLock && !hasNpmLock) {
+    console.log("Using yarn as package manager, as there is a yarn.lock file.");
+    return YarnPackageManager;
+  } else if (hasNpm && hasNpmLock && !hasYarnLock) {
+    console.log(
+      "Using npm as package manager, as there is a package-lock.json file."
```
Comment (threepointone) [inline packages/wrangler/src/package-manager.ts:21]: Opinion - we should consider npm as a "default", and not log a message when it's the choice.

File: packages/wrangler/src/package-manager.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    );
+    return NpmPackageManager;
+  } else if (hasYarn && !hasNpm) {
+    console.log(
+      "Using yarn as package manager, as there is no npm binary on the PATH."
```
Comment (threepointone) [inline packages/wrangler/src/package-manager.ts:26]: Opinion - We probably shouldn't surface the reason here to the user (it'll make them concerned that they do need to install npm)

File: packages/wrangler/src/package-manager.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    );
+    return YarnPackageManager;
+  } else if (hasNpm && !hasYarn) {
+    console.log(
+      "Using npm as package manager, as there is no yarn binary on the PATH."
```
Comment (threepointone) [inline packages/wrangler/src/package-manager.ts:31]: Opinion - As above, we should consider npm as a default, but also we shouldn't surface the _lack_ of something to the user as reasoning.

**Review by threepointone — APPROVED** [review]
This is looking very good to me! I dropped some _opinions_, feel free to take a call on them. 

We should definitely test this after we land it.

File: packages/wrangler/src/package-manager.ts (hunk 5 above)
Comment (petebacondarwin) [reply packages/wrangler/src/package-manager.ts]: Funnily enough, when I was playing with this locally, I came to the same conclusions.

**Comment by petebacondarwin** [comment]
I reworked the console messages to be less worrying. The only "warning" now is when there is a lock-file but no equivalent package manager in the PATH.

File: packages/wrangler/src/package-manager.ts (hunk 5 above)
Comment (threepointone) [reply packages/wrangler/src/package-manager.ts]: image

**Pull request merged**